File: day-16/main.rb
# Advent of Code 2015 - Day 16

class Aunt
  AUNT_REGEXP = /^Sue (\d+): (.*)$/
  INFO_REGEXP = /(\w+): (\d+)/

  attr_reader :number
  attr_reader :infos

  def self.from_line(line)
    number, infos = line.match(AUNT_REGEXP)[1..2]
    infos = Hash[infos.scan(INFO_REGEXP).map { |k, v| [k.to_sym, v.to_i] }]
    new(number, infos)
  end

  def initialize(number, infos)
    @number = number
    @infos = infos
  end

  def match?(wanted, relop = {})
    infos.each do |key, value|
      next unless wanted.key?(key)
      return false unless value_match?(value, wanted[key], relop[key])
    end
    true
  end

  private

  def value_match?(aunt_value, wanted_value, relop)
    case relop
    when :gt
      aunt_value > wanted_value
    when :lt
      aunt_value < wanted_value
    else
      aunt_value == wanted_value
    end
  end
end

module Day16
  WANTED = {
    :children    => 3,
    :cats        => 7,
    :samoyeds    => 2,
    :pomeranians => 3,
    :akitas      => 0,
    :vizslas     => 0,
    :goldfish    => 5,
    :trees       => 3,
    :cars        => 2,
    :perfumes    => 1,
  }.freeze

  WANTED_RELOP = {
    :cats        => :gt,
    :trees       => :gt,
    :pomeranians => :lt,
    :goldfish    => :lt,
  }.freeze

  def self.read_input
    File.read(File.expand_path("../input.txt", __FILE__))
  end

  def self.find_aunt(aunts, wanted, relop = {})
    wanted_aunts = aunts.select { |aunt| aunt.match?(wanted, relop) }
    raise "Failed to find aunt that matches criteria." if wanted_aunts.empty?
    raise "Failed to find uniquely matching aunt." if wanted_aunts.size != 1
    wanted_aunts.first
  end

  def self.main
    aunts = []

    lines = read_input.split("\n").reject(&:empty?)
    lines.each { |line| aunts << Aunt.from_line(line) }

    puts "Part One: #{find_aunt(aunts, WANTED).number}"
    puts "Part Two: #{find_aunt(aunts, WANTED, WANTED_RELOP).number}"
  end
end

Day16.main if "main" == to_s

File: day-16/main.py
"""Advent of Code 2015 - Day 16."""

import operator
import re
from pathlib import Path


def read_input():
    """Read input file and split into individual lines returned as a list."""
    return Path(__file__).with_name('input.txt').read_text().splitlines()


class Aunt:
    """Aunt identified by a number and described by various information."""

    # Regular expression for extracting number and description of aunt.
    AUNT_REGEXP = re.compile(r'^Sue (\d+): (.*)$')

    # Regular expression for further processing an aunt description.
    INFO_REGEXP = re.compile(r'(\w+): (\d+)')

    @classmethod
    def from_line(cls, line):
        """Initialize an aunt from a line from the input file."""
        number, infos = cls.AUNT_REGEXP.search(line).groups()
        return cls(int(number),
                   [(k, int(v)) for k, v in cls.INFO_REGEXP.findall(infos)])

    def __init__(self, number, infos):
        """Initialize an aunt with its number and a description."""
        self.number = number
        self.infos = infos

    def match(self, wanted_value, wanted_relop):
        """Check if the aunt matches the given description."""
        for key, value in self.infos:
            relop = wanted_relop.get(key, operator.eq)
            if not relop(value, wanted_value[key]):
                return False
        return True


def find_aunt(aunts, wanted_value, wanted_relop={}):
    """Find aunt that uniquely matches the given description."""
    matcher = operator.methodcaller('match', wanted_value, wanted_relop)
    matched = list(filter(matcher, aunts))
    if len(matched) == 1:
        return matched[0]
    raise RuntimeError('Failed to find uniquely matching aunt.')


def main():
    """Main entry point of puzzle solution."""
    WANTED_VALUE = {
        'children': 3,
        'cats': 7,
        'samoyeds': 2,
        'pomeranians': 3,
        'akitas': 0,
        'vizslas': 0,
        'goldfish': 5,
        'trees': 3,
        'cars': 2,
        'perfumes': 1,
    }

    WANTED_RELOP = {
        'cats': operator.gt,
        'trees': operator.gt,
        'pomeranians': operator.lt,
        'goldfish': operator.lt,
    }

    aunts = [Aunt.from_line(line) for line in read_input()]

    part_one = find_aunt(aunts, WANTED_VALUE).number
    part_two = find_aunt(aunts, WANTED_VALUE, WANTED_RELOP).number

    print('Part One: {}'.format(part_one))
    print('Part Two: {}'.format(part_two))


if __name__ == '__main__':
    main()
